src/AzureExtension/DevBox/DevBoxInstance.cs

```diff
@@ -541,6 +609,109 @@ public IAsyncOperation<ComputeSystemOperationResult> ModifyPropertiesAsync(strin
+            _log.Error($"Error getting thumbnail for {DisplayName}");
+            return new ComputeSystemOperationResult(new NotSupportedException(), Resources.GetResource(Constants.DevBoxUnableToPerformOperationKey), errorString);
+        }).AsAsyncOperation();
+    }
+
+    public IAsyncOperation<ComputeSystemOperationResult> PinToStartMenuAsync()
+    {
+        return DoPinActionAsync("startMenu", "pin");
+    }
+
+    public IAsyncOperation<ComputeSystemOperationResult> UnpinFromStartMenuAsync()
+    {
+        return DoPinActionAsync("startMenu", "unpin");
+    }
+
+    public IAsyncOperation<ComputeSystemOperationResult> PinToTaskbarAsync()
+    {
+        return DoPinActionAsync("taskbar", "pin");
+    }
+
+    public IAsyncOperation<ComputeSystemOperationResult> UnpinFromTaskbarAsync()
+    {
+        return DoPinActionAsync("taskbar", "unpin");
+    }
+
+    public IAsyncOperation<ComputeSystemPinnedResult> GetPinStatusAsync(string location)
+    {
+        return Task.Run(() =>
+        {
+            var exitcode = ExitCodeInvalid;
+            var psi = new ProcessStartInfo();
+            psi.UseShellExecute = true;
+            psi.FileName = string.Format(CultureInfo.InvariantCulture, ProtocolPinString, location, "status", WorkspaceId, DisplayName, Username, Environment);
+            Process? p = Process.Start(psi);
+            if (p != null)
+            {
+                p.WaitForExit();
+                exitcode = p.ExitCode;
+                if (exitcode == ExitCodePinned)
+                {
+                    return new ComputeSystemPinnedResult(true);
+                }
+                else if (exitcode == ExitCodeUnpinned)
+                {
+                    return new ComputeSystemPinnedResult(false);
+                }
+            }
+
+            var errorString = $"GetPinStatusAsync from location {location} failed with exitcode: {exitcode}";
+            _log.Error($"Error getting thumbnail for {DisplayName}");
+            return new ComputeSystemPinnedResult(new NotSupportedException(), Resources.GetResource(Constants.DevBoxUnableToPerformOperationKey), errorString);
+        }).AsAsyncOperation();
+    }
+
+    public IAsyncOperation<ComputeSystemPinnedResult> GetIsPinnedToStartMenuAsync()
+    {
+        return GetPinStatusAsync("startmenu");
+    }
+
+    public IAsyncOperation<ComputeSystemPinnedResult> GetIsPinnedToTaskbarAsync()
+    {
+        return GetPinStatusAsync("taskbar");
+    }
+
+    public async void LoadWindowsAppParameters()
+    {
+        if (RemoteConnectionData is null)
+        {
+            await GetRemoteLaunchURIsAsync(new Uri(DevBoxState.Uri));
+        }
+
+        var uriString = RemoteConnectionData?.CloudPcConnectionUrl;
+        if (uriString == null)
+        {
+            return;
+        }
+
+        var launchUri = new Uri(uriString);
+        WorkspaceId = HttpUtility.ParseQueryString(launchUri.Query)["cpcid"];

```

Comment:
Will `ParseQueryString` throw if the key does not exist?

Reply:
I tested it and it just returns null.  But I've gone ahead and put try/catch around there anyways because the Uri can throw.